### PR TITLE
treat ignored addons as 'locked'

### PR DESCRIPTION
**Pull request by torkus**
A yak that needs shaving prior to implementing mutual dependency handing.

- [x] ~rename `ignore?` to `locked`~
    - too much effort. instead, I'm going to treat ignored addons as if they are locked 
- [x] ignored/locked addons cannot be uninstalled
- [x] if any part of a bundle is ignored, the whole addon is ignored and cannot be uninstalled
- [x] gui, ignored/locked addons prevent bulk uninstallation operations
- [x] an addon cannot be installed if it conflicts with another ignored/locked addon
- [x] gui, an addon can be explicitly unlocked
- [ ] review

### CODE REVIEW

File: src/strongbox/addon.clj [hunk 1]
```diff
@@ -29,15 +29,30 @@
 
 (defn-spec remove-addon nil?
   "removes the given addon. if addon is part of a group, all addons in group are removed"
-  [addon-dir ::sp/addon-dir, toc :addon/toc]
-  (if (contains? toc :group-addons)
-    ;; top-level toc is contained in the :group-addons list
-    (doseq [subtoc (:group-addons toc)]
-      (-remove-addon addon-dir (:dirname subtoc)))
-    (-remove-addon addon-dir (:dirname toc))))
+  [addon-dir ::sp/addon-dir, addon :addon/installed]
```
Comment (torkus) [inline src/strongbox/addon.clj:32]: review usage of `addon/installed` - is it more or less permissive than `toc` ? if less permissive, then revert.

File: src/strongbox/addon.clj [hunk 2]
```diff
@@ -61,13 +70,18 @@
                          next-best (first addons)
                          new-data {:group-addons addons
                                    :group-addon-count (count addons)}
-                         next-best-label (-> next-best :group-id fs/base-name)]
+                         next-best-label (-> next-best :group-id fs/base-name)
+                         ;; add a group-level ignore flag if any bundled addon is being ignored
+                         ;; todo: test for this
```
Comment (torkus) [inline src/strongbox/addon.clj:75]: todo

File: src/strongbox/addon.clj [hunk 3]
```diff
@@ -165,7 +182,26 @@
...
+    (info (format "installing '%s' version '%s'" (:label addon) (:version addon)))
     (install-addon)
-    (let [retval (update-nfo-files)]
-      (info (:label addon) "installed.")
-      retval)))
+    (update-nfo-files)))
+
+;;
+
+(defn-spec ignored-dir-list (s/coll-of string?)
```
Comment (torkus) [inline src/strongbox/addon.clj:192]: docstr

Comment (torkus) [reply src/strongbox/addon.clj:192]: s/coll-of ::sp/dirname

File: src/strongbox/addon.clj [hunk 4]
```diff
@@ -165,7 +182,26 @@
...
+  [addon-list (s/nilable :addon/installed-list)]
+  (->> addon-list (filterv :ignore?) (map :group-addons) flatten (map :dirname) (remove nil?) set))
```
Comment (torkus) [inline src/strongbox/addon.clj:194]: just filter

File: src/strongbox/addon.clj [hunk 5]
```diff
@@ -165,7 +182,26 @@
...
+
+(defn-spec overwrites-ignored? boolean?
+  "returns true if given archive file would unpack over *any* ignored addon.
+  this includes already installed versions of itself and is another check against modifying ignored addons."
+  [downloaded-file ::sp/archive-file, addon-list (s/nilable :addon/installed-list)]
```
Comment (torkus) [inline src/strongbox/addon.clj:199]: can we get away with a list of dirnames instead of installed-list?

File: src/strongbox/catalogue.clj [hunk 6]
```diff
@@ -15,25 +15,25 @@
     [wowinterface-api :as wowinterface-api]
     [github-api :as github-api]]))
 
-(defn expand-summary
-  [addon-summary game-track]
+(defn-spec expand-summary (s/or :ok (s/merge :addon/expandable :addon/source-updates), :error nil?)
```
Comment (torkus) [inline src/strongbox/catalogue.clj:18]: docstr

File: src/strongbox/core.clj [hunk 7]
```diff
@@ -409,44 +409,52 @@
...
 (defn-spec install-addon-guard (s/or :ok (s/coll-of ::sp/extant-file), :passed-tests true?, :error nil?)
-  "downloads an addon and installs it. handles http and non-http errors, bad zip files, bad addons"
+  "downloads an addon and installs it. 
+  handles http and non-http errors, bad zip files, bad addons, bad directories."
+  ([addon :addon/installable]
+   (install-addon-guard addon (selected-addon-dir)))
   ([addon :addon/installable, install-dir ::sp/extant-dir]
    (install-addon-guard addon install-dir false))
   ([addon :addon/installable, install-dir ::sp/extant-dir, test-only? boolean?]
+   ;;(info (format "preparing '%s' version '%s'" (:label addon) (:version addon)))
```
Comment (torkus) [inline src/strongbox/core.clj:419]: rm

File: src/strongbox/core.clj [hunk 8]
```diff
@@ -409,44 +409,52 @@
...
      ;; do some pre-installation checks
-     (:ignore? addon) (warn "failing to install addon, addon is being ignored:" install-dir)
-     (not (fs/writeable? install-dir)) (error "failing to install addon, directory not writeable:" install-dir)
+     (:ignore? addon) (error "refusing to install addon, addon is being ignored:" (:name addon))
+     (not (fs/writeable? install-dir)) (error "refusing to install addon, directory not writeable:" install-dir)
 
      :else ;; attempt downloading and installing addon
 
-     (let [downloaded-file (download-addon addon install-dir)
+     (let [
```
Comment (torkus) [inline src/strongbox/core.clj:427]: whitespace

File: src/strongbox/core.clj [hunk 9]
```diff
@@ -945,8 +954,10 @@
 
 ;;
 
-(defn-spec -upgrade-nfo nil?
-  "given an addon, upgrades it's nfo file to the current nfo spec."
+(defn-spec -upgrade-nfo-files-to-v2 nil?
```
Comment (torkus) [inline src/strongbox/core.clj:957]: singular

File: src/strongbox/core.clj [hunk 10]
```diff
@@ -1075,6 +1088,18 @@
   (-> (get-state) :selected-installed vec remove-many-addons)
   nil)
 
+(defn-spec ignore-selected nil?
```
Comment (torkus) [inline src/strongbox/core.clj:1091]: docstr

File: src/strongbox/core.clj [hunk 11]
```diff
@@ -1075,6 +1088,18 @@
...
+  []
+  (->> (get-state) :selected-installed :dirname (mapv (partial nfo/ignore (selected-addon-dir))))
+  (refresh))
+
+(defn-spec clear-ignore-selected nil?
```
Comment (torkus) [inline src/strongbox/core.clj:1096]: docstr

File: src/strongbox/nfo.clj [hunk 12]
```diff
@@ -62,15 +62,17 @@
...
   (let [path (nfo-path install-dir addon-dirname)]
     (utils/dump-json-file path (derive addon primary? game-track))
     path))
 
-(defn-spec upgrade-nfo ::sp/extant-file
-  "refreshes the nfo data for the given addon. does NOT bump installed version"
+(defn-spec upgrade-nfo-to-v2 (s/or :ok ::sp/extant-file, :invalid-updates nil?)
+  "upgrades the nfo data from v1 nfo ('anything') to v2, if possible. 
+  requires an addon directory and an addon version as an absolute minimum.
+  returns nil if insufficient/invalid data."
```
Comment (torkus) [inline src/strongbox/nfo.clj:75]: ticks on nil

File: src/strongbox/nfo.clj [hunk 13]
```diff
@@ -139,6 +141,44 @@
        (s/valid? :addon/nfo (utils/load-json-file-safely (nfo-path install-dir (:dirname addon))
                                                          {:transform-map {:installed-game-track keyword}}))))
 
+(defn-spec update-nfo nil?
```
Comment (torkus) [inline src/strongbox/nfo.clj:144]: tests

File: src/strongbox/nfo.clj [hunk 14]
```diff
@@ -139,6 +141,44 @@
...
+  "updates *existing* nfo data with new values.
+  content is only written if valid."
```
Comment (torkus) [inline src/strongbox/nfo.clj:146]: how is this different from 'upgrade-nfo' ?

File: src/strongbox/nfo.clj [hunk 15]
```diff
@@ -139,6 +141,44 @@
...
+  [install-dir ::sp/extant-dir, dirname ::sp/dirname, updates map?]
+  (let [path (nfo-path install-dir dirname)
+        nfo (read-nfo install-dir dirname)
+        new-nfo (merge nfo updates)
+        ;; convenience for implicitly ignored addons.
+        ;; if the nfo file was missing and the new request is to remove the flag, set it to false instead
```
Comment (torkus) [inline src/strongbox/nfo.clj:152]: elaborate on this a bit more. the logic is trickier than I'm leading on

File: src/strongbox/nfo.clj [hunk 16]
```diff
@@ -139,6 +141,44 @@
...
+        new-nfo (if (and (not (fs/exists? path))
+                         (= {:ignore? nil} new-nfo))
+                  {:ignore? false}
+                  new-nfo)
+        ]
+    (if (= {:ignore? nil} new-nfo)
+      ;; edge case. a valid nfo file is also simply a `ignore: True`
+      ;; when this is true, just delete the nfo file
+      (rm-nfo path)
+      (when (s/valid? :addon/nfo new-nfo)
```
Comment (torkus) [inline src/strongbox/nfo.clj:162]: make this an `if` with an error case. I want to know, even if it's in debug, if updating the nfo data fails

File: src/strongbox/nfo.clj [hunk 17]
```diff
@@ -139,6 +141,44 @@
...
+        (utils/dump-json-file path new-nfo))))
+  nil)
+
+(defn-spec ignore nil?
+  "explicitly ignores this addon by setting the `ignore?` flag to `true`.
+  Prevents any changes made by strongbox to that addon."
```
Comment (torkus) [inline src/strongbox/nfo.clj:168]: "Prevent any changes made by strongbox to this addon."

File: src/strongbox/nfo.clj [hunk 18]
```diff
@@ -139,6 +141,44 @@
...
+  [install-dir ::sp/extant-dir, dirname ::sp/dirname]
+  (update-nfo install-dir dirname {:ignore? true})
+  nil)
+
+(defn-spec clear-ignore nil?
```
Comment (torkus) [inline src/strongbox/nfo.clj:173]: test

File: src/strongbox/nfo.clj [hunk 19]
```diff
@@ -139,6 +141,44 @@
...
+        ]
+    (if (= {:ignore? nil} new-nfo)
+      ;; edge case. a valid nfo file is also simply a `ignore: True`
+      ;; when this is true, just delete the nfo file
+      (rm-nfo path)
+      (when (s/valid? :addon/nfo new-nfo)
+        (utils/dump-json-file path new-nfo))))
+  nil)
+
+(defn-spec ignore nil?
```
Comment (torkus) [inline src/strongbox/nfo.clj:166]: test

File: src/strongbox/nfo.clj [hunk 20]
```diff
@@ -139,6 +141,44 @@
...
+  "explicitly ignores this addon by setting the `ignore?` flag to `true`.
+  Prevents any changes made by strongbox to that addon."
+  [install-dir ::sp/extant-dir, dirname ::sp/dirname]
+  (update-nfo install-dir dirname {:ignore? true})
+  nil)
```
Comment (torkus) [inline src/strongbox/nfo.clj:171]: no need for an explicit `nil` here, `update-nfo` returns nil already

File: src/strongbox/specs.clj [hunk 21]
```diff
@@ -156,7 +156,9 @@
                                     :opt [::ignore?])))
 
 ;; minimum amount of data required to create a nfo file. the rest is derived.
-(s/def :addon/nfo-input-minimum (s/keys :req-un [::version ::name ::url :addon/source :addon/source-id]))
+(s/def :addon/nfo-input-minimum (s/keys :req-un [::version ::name
```
Comment (torkus) [inline src/strongbox/specs.clj:159]: add note 'intermediate spec'

File: src/strongbox/ui/gui.clj [hunk 22]
```diff
@@ -262,21 +262,27 @@
...
-                            :resizable? false
-                            :type :warning
-                            :option-type :ok-cancel
-                            :default-option :no
-                            :success-fn (async-handler core/remove-selected))]
-      (-> dialog ss/pack! ss/show!)
-      nil)))
+    (if (utils/any (mapv :ignore? selected))
+      (-> (ss/dialog :parent (select-ui :#root)
+                     :content "Selection contains ignored addons. Unlock them first, then delete."
```
Comment (torkus) [inline src/strongbox/ui/gui.clj:267]: 'Unlock them first' -> 'Un-ignore them first' (?)

File: src/strongbox/ui/gui.clj [hunk 23]
```diff
@@ -918,7 +927,11 @@
                                 (let [exit-ev (java.awt.event.WindowEvent. newui java.awt.event.WindowEvent/WINDOW_CLOSING)]
                                   (.dispatchEvent newui exit-ev))))]
 
-        view-menu (build-theme-menu)
+        view-menu (into [(ss/action :name "Refresh" :key "F5" :handler (async-handler core/refresh))
+                         :separator
+                         ]
+
```
Comment (torkus) [inline src/strongbox/ui/gui.clj:933]: whitespace

File: src/strongbox/zip.clj [hunk 24]
```diff
@@ -115,19 +115,26 @@
 ;;
 ;;
 
+(defn top-level-directories
```
Comment (torkus) [inline src/strongbox/zip.clj:118]: defn-spec
docstr

File: src/strongbox/zip.clj [hunk 25]
```diff
@@ -115,19 +115,26 @@
...
+  [zipfile-entries]
+  (filter (every-pred :dir? :toplevel?) zipfile-entries))
+
+(defn top-level-files
```
Comment (torkus) [inline src/strongbox/zip.clj:122]: defn-spec
docstr

File: test/strongbox/catalogue_test.clj [hunk 26]
```diff
@@ -209,3 +209,22 @@
 
     (testing "catalogue with an incorrect total yields `nil`"
       (is (nil? (catalogue/validate catalogue-with-incorrect-total))))))
+
+;;
+
+;; todo: 404 and 500 errors need better support than "unhandled exception"
```
Comment (torkus) [inline test/strongbox/catalogue_test.clj:215]: todo

File: test/strongbox/core_test.clj [hunk 27]
```diff
@@ -593,18 +593,45 @@
...
-        (is (= ["BundledAddon" "EveryAddon"] directory-list))
-        (is (= expected-nfo (nfo/read-nfo-file install-dir "BundledAddon")))))))
+        (is (= ["EveryAddon" "EveryAddon-BundledAddon"] directory-list))
+        (is (= expected-nfo (nfo/read-nfo-file install-dir "EveryAddon-BundledAddon")))))))
+
+(deftest install-bundled-addon-overwriting-ignored-addon
+  (testing "installing/unzipping an addon with a shared mutual dependency of an addon that is ignored isn't possible"
+    (with-running-app
+      (let [install-dir (helper/addons-path)
+            _ (core/set-addon-dir! install-dir)
```
Comment (torkus) [inline test/strongbox/core_test.clj:603]: this pattern is getting tiresome. have `helper/addons-path` set the addon dir at the same time as well

File: test/strongbox/core_test.clj [hunk 28]
```diff
@@ -593,18 +593,45 @@
...
+            install-dir-contents #(->> install-dir fs/list-dir (map fs/base-name) sort)
```
Comment (torkus) [inline test/strongbox/core_test.clj:604]: I use this often enough it should probably join help/addons-path in test_helper.clj

File: test/strongbox/core_test.clj [hunk 29]
```diff
@@ -593,18 +593,45 @@
...
+            addon {:name "everyaddon" :label "EveryAddon" :version "0.1.2" :url "https://group.id/never/fetched"
+                   :source "curseforge" :source-id 1
+                   :-testing-zipfile (fixture-path "everyaddon--0-1-2.zip")}]
+
+        (core/install-addon addon)
+        (is (= ["EveryAddon" "EveryAddon-BundledAddon"] (install-dir-contents)))
+
+        (fs/mkdir (utils/join install-dir "EveryAddon" ".git"))
+        (core/load-installed-addons) ;; refresh our knowledge of what is installed essentially
```
Comment (torkus) [inline test/strongbox/core_test.clj:613]: rm 'essentially'

File: src/strongbox/addon.clj (hunk 5 above)
Comment (torkus) [reply src/strongbox/addon.clj:199]: too late

File: src/strongbox/core.clj [hunk 30]
```diff
@@ -1000,6 +1011,9 @@
    {:when (get-state :profile?)}
 
    ;; rename any occurances of `.wowman.json` to `.strongbox.json`
+   ;; todo: bug here. files are migrated across, but the data may be invalid for nfo v2.
```
Comment (torkus) [inline src/strongbox/core.clj:1014]: fix

Comment (torkus) [reply src/strongbox/core.clj:1014]: this will get it's own PR
